src/mbgl/renderer/layers/render_circle_layer.cpp

```diff
@@ -308,20 +333,58 @@ void RenderCircleLayer::update(gfx::ShaderRegistry& shaders,
 
+    const auto& evaluated = static_cast<const CircleLayerProperties&>(*evaluatedProperties).evaluated;
+
     for (const RenderTile& tile : *renderTiles) {
         const auto& tileID = tile.getOverscaledTileID();
 
-        if (tileLayerGroup->getDrawableCount(renderPass, tileID) > 0) {
-            continue;
-        }
-
         const LayerRenderData* renderData = getRenderDataForPass(tile, renderPass);
         if (!renderData) {
             removeTile(renderPass, tileID);
             continue;
         }
 
         auto& bucket = static_cast<CircleBucket&>(*renderData->bucket);
+        const auto& paintPropertyBinders = bucket.paintPropertyBinders.at(getID());
+
+        const CircleInterpolateUBO interpolateUBO = {
+            /* .color_t = */ std::get<0>(paintPropertyBinders.get<CircleColor>()->interpolationFactor(state.getZoom())),
+            /* .radius_t = */
+            std::get<0>(paintPropertyBinders.get<CircleRadius>()->interpolationFactor(state.getZoom())),
+            /* .blur_t = */ std::get<0>(paintPropertyBinders.get<CircleBlur>()->interpolationFactor(state.getZoom())),
+            /* .opacity_t = */
+            std::get<0>(paintPropertyBinders.get<CircleOpacity>()->interpolationFactor(state.getZoom())),
+            /* .stroke_color_t = */
+            std::get<0>(paintPropertyBinders.get<CircleStrokeColor>()->interpolationFactor(state.getZoom())),
+            /* .stroke_width_t = */
+            std::get<0>(paintPropertyBinders.get<CircleStrokeWidth>()->interpolationFactor(state.getZoom())),
+            /* .stroke_opacity_t = */
+            std::get<0>(paintPropertyBinders.get<CircleStrokeOpacity>()->interpolationFactor(state.getZoom())),
+            /* .padding = */ 0};
+
+        tileLayerGroup->observeDrawables(renderPass, tileID, [&](gfx::Drawable& drawable) {
+            drawable.mutableUniformBuffers().createOrUpdate(CircleInterpolateUBOName, &interpolateUBO, context);
+        });
+
+        if (tileLayerGroup->getDrawableCount(renderPass, tileID) > 0) {
+            continue;
+        }
+
+        circleVertexAttrs.clear();
+
+        auto propertiesAsUniforms = circleVertexAttrs.readDataDrivenPaintProperties<CircleColor,
+                                                                                    CircleRadius,
+                                                                                    CircleBlur,
+                                                                                    CircleOpacity,
+                                                                                    CircleStrokeColor,
+                                                                                    CircleStrokeWidth,
+                                                                                    CircleStrokeOpacity>(
+            paintPropertyBinders, evaluated);
+
+        auto circleShader = circleShaderGroup->getOrCreateShader(context, propertiesAsUniforms);
+        if (!circleShader) {
```

Comment:
This is the code where our vertex attributes reads from data driven paint properties and generates the permutation as return to be used to create the shader variant.